gateway/broker/session.go

```diff
@@ -0,0 +1,304 @@
+			return wsConn.WriteMessage(websocket.BinaryMessage, data)
+		}
+	case "tcp":
+		if tcpConn, ok := s.Agent.Connection.(net.Conn); ok {
+			_, err := tcpConn.Write(data)
+			return err
+		}
+	default:
+		return nil
+	}
+	return nil
+}
+
+func (s *Session) ReadFromAgent() (int, []byte, error) {
+	switch s.Agent.ConnType {
+	case "websocket":
+		if wsConn, ok := s.Agent.Connection.(*websocket.Conn); ok {
+			_, message, err := wsConn.ReadMessage()
+			return len(message), message, err
+		}
+	case "tcp":
+		if tcpConn, ok := s.Agent.Connection.(net.Conn); ok {
+			buffer := make([]byte, 16*1024)
+			n, err := tcpConn.Read(buffer)
+			if err != nil {
+				return 0, nil, err
+			}
+			return n, buffer, nil
+		}
+	default:
+		return 0, nil, nil
+	}
+	return 0, nil, nil
+}
+
+func (s *Session) Close() {
+	close(s.dataChannel)
+	if s.Consumer != nil {
+		if conn, ok := s.Consumer.Connection.(net.Conn); ok {
+			conn.Close()
+		}
+	}
+	if s.Agent != nil {
+		switch s.Agent.ConnType {
+		case "websocket":
+			if wsConn, ok := s.Agent.Connection.(*websocket.Conn); ok {
+				wsConn.Close()
+			}
+		case "tcp":
+			if tcpConn, ok := s.Agent.Connection.(net.Conn); ok {
+				tcpConn.Close()
+			}
+		}
+	}
+	BrokerInstance.sessions.Delete(s.ID)
+	log.Infof("Session %s closed", s.ID)
+}
+
+// forward data from agent to tcp
+func (s *Session) ForwardToTCP(data []byte) {
+	log.Infof("Forwarded %d bytes to TCP session %s", len(data), s.ID)
+	s.dataChannel <- data
+}
+
+// this will spam data from tcp to agent wsconn
+func (s *Session) StartingForwardind(data []byte) error {
+
+	header := &Header{
+		SID: s.ID,
+		Len: uint32(len(data)),
+	}
+
+	framedData := append(header.Encode(), data...)
+
+	if err := s.SendToAgent(framedData); err != nil {
+		log.Infof("Failed to send first RDP packet: %v", err)
+		return nil
+	}
+
+	log.Infof("Sent first RDP packet: %d bytes", len(data))
+
+	// Continue reading from TCP connection (not from agent!)
+	if tcpConn, ok := s.Consumer.Connection.(net.Conn); ok {
+		buffer := make([]byte, 16*1024)
+		for {
+			n, err := tcpConn.Read(buffer)
+			if err != nil {
+				if err != io.EOF {
+					log.Infof("TCP read error: %v", err)
+				}
+				break
+			}
+
+			if n > 0 {
+				log.Infof("TCP -> Agent: %d bytes for session %s", n, s.ID)
+
+				header := &Header{
+					SID: s.ID,
+					Len: uint32(n),
+				}
+				framedData := append(header.Encode(), buffer[:n]...)
+
+				if err := s.SendToAgent(framedData); err != nil {
+					log.Infof("Failed to send RDP data to agent: %v", err)
+					break
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// this will forward data from agent to tcp
+func (s *Session) SendAgentToTCP() {
+	for data := range s.dataChannel {
+		log.Infof("Agent -> TCP: %d bytes for session %s", len(data), s.ID)
+
+		conn, _ := s.Consumer.Connection.(net.Conn)
+		if _, err := conn.Write(data); err != nil {
+			log.Infof("TCP write error: %v", err)
+			break
+		}
+	}
+
+}
+
+func CreateRDPSession(
```

Comment:
this functoin is to specific for rdp protocol, we can think in a way to create a abstraction to remove trom this layer